platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -116,35 +114,66 @@ public class HalfDiskHashMap implements AutoCloseable, Snapshotable, FileStatist
     private final ReusableBucketPool bucketPool;
     /** Store for session data during a writing transaction */
     private IntObjectHashMap<BucketMutation> oneTransactionsData = null;
+
+    // Fields related to flushes
+
     /**
      * The thread that called startWriting. We use it to check that other writing calls are done on
      * same thread
      */
     private Thread writingThread;
 
-    /** Executor for parallel bucket reads/updates in {@link #endWriting()} */
-    private static volatile ExecutorService flushExecutor = null;
+    private final AtomicReference<SubmitTask> currentSubmitTask = new AtomicReference<>();
+
+    /** Number of buckets updated during flush */
+    private final AtomicInteger numBuckets = new AtomicInteger();
+
+    /**
+     * Number of bucket tasks that can be scheduled at the moment, i.e. MAX_IN_FLIGHT minus
+     * the number of buckets currently being processed
+     */
+    private final AtomicInteger bucketPermits = new AtomicInteger(MAX_IN_FLIGHT);
+
+    /**
+     * The last created "store bucket" task, or null if no such tasks have been created yet in
+     * the current flush. This reference is used to set store task dependencies to make sure
+     * only a single bucket is written to disk at a time
+     */
+    private final AtomicReference<StoreBucketTask> lastStoreTask = new AtomicReference<>();
+
+    /**
+     * Number of "store bucket" tasks created so far in the current flush. This counter is
+     * compared against {@link #numBuckets} to manage the first and the last "store bucket"
+     * task dependencies
+     */
+    private final AtomicInteger storeBucketTasksCreated = new AtomicInteger();
+
+    /**
+     * A task used in {@link #endWriting()} to wait for all submitted tasks to complete. It
+     * is set as a "next" dependency for the last created "store bucket" task.
+     */
+    private final AtomicReference<AbstractTask> notifyTaskRef = new AtomicReference<>();
```

Comment:
Does not have to be atomic or even volatile.

Reply:
When an exception occurs, this task is notified to complete exceptionally. It happens without any changes to bucketPermits. I'd leave `notifyTaskRef` an atomic reference.

Reply:
`notifyTaskRef` is set before any task that might complete it exceptionally is created and passed to another thread for execution. There is enough "happens-before" to ensure its visibility.